### PR TITLE
`weak_ptr` conversions can sometimes avoid locking

### LINKED ISSUE #258: <memory>: weak_ptr's converting constructors could sometimes avoid locking
`weak_ptr`'s converting copy/move constructors have a subtle bit of code:

https://github.com/microsoft/STL/blob/d0d724fb916c59f1c4f2830330ed967a72d2d2b8/stl/inc/memory#L1622-L1625

https://github.com/microsoft/STL/blob/d0d724fb916c59f1c4f2830330ed967a72d2d2b8/stl/inc/memory#L1631-L1635

They `lock()` to get a (possibly empty) `shared_ptr` before performing the converting construction. This means that we're paying an expensive "increment refcount if nonzero" and a decrement.

We're doing this for a real reason (not commented in the source code, but covered by tests). These ctors are performing a raw pointer conversion from `Y *` to `T *`. That might perform a pointer adjustment to traverse an inheritance hierarchy. But because this is `weak_ptr`, the pointed-to object might have expired. Usually that doesn't matter - the offset for the pointer adjustment is known statically. But for **virtual inheritance**, the offset depends on the most derived type, and that requires the object to be alive! (The wording for the Core Language says this in a more obscure way.) So we lock to ensure that either the object remains alive during the conversion, or it's expired and we'll get an empty `weak_ptr`.

That's great, except that it's too aggressive. It locks when converting from `X *` to `const X *` which never requires a pointer adjustment (and is easy to detect). It also locks when the conversion doesn't involve a virtual base, but I don't know whether we can detect that via type traits (and it's unlikely to matter in practice).

Now that we have `if constexpr`, we could easily optimize this.

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Resolves #258

Test happen to recreate the issue for me on attempt #0
if unconditionally avoid locking

### CODE REVIEW

File: stl/inc/memory [hunk 1]
```diff
@@ -2949,6 +2949,18 @@ _NODISCARD enable_if_t<is_unbounded_array_v<_Ty>, shared_ptr<_Ty>> allocate_shar
 
 template <class _Ty>
 class weak_ptr : public _Ptr_base<_Ty> { // class for pointer to reference counted resource
+private:
+    template <class _Ty2,
```
Comment (miscco) [inline stl/inc/memory:2953]: Stupid question, but why isnt this an inline template variable that defaults to false?

File: tests/std/tests/Dev10_851347_weak_ptr_virtual_inheritance/test.cpp [hunk 2]
```diff
@@ -60,6 +62,52 @@ void test_owner_equal() {
...
+
+    struct base3 {
+        int k = 0;
+    };
+
+    struct derived : virtual base1, virtual base2, base3 {};
+
+    for (int i = 0; i < 10; ++i) {
+        // not make_shared -- with make_shared the test would not catch errors
+        shared_ptr<derived> d{new derived{}};
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply tests/std/tests/Dev10_851347_weak_ptr_virtual_inheritance/test.cpp:82]: * There's no such CTAD rule for `shared_ptr` that deduces it from taw pointer
 * This test covers C++14 as well
 * For this test it is important where `derived` and where `base...` type is used, so I'd prefer it all explicit anyway

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
@AlexGuteniev @CaseyCarter I've pushed changes to fix test failures for `/clr:pure`, the Betrayer of Hope - the simplest thing is to entirely disable this optimization.

**Comment by CaseyCarter** [comment]
> @AlexGuteniev @CaseyCarter I've pushed changes to fix test failures for `/clr:pure`, the Betrayer of Hope - the simplest thing is to entirely disable this optimization.

Having recently begun my fourth trip through the Wheel of Time, I'm deeply offended that you would liken `/clr:pure` to Ishamael. Not even one of the Forsaken is as dark and twisted as `/clr:pure`!

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for implementing this long-dreamed-of optimization! :cat2: :zzz: :heart_eyes_cat:

**Pull request merged**